src/XR/features/WebXRWalkingLocomotion.ts

```diff
@@ -0,0 +1,438 @@
+    private _movement: Vector3 = new Vector3();
+    private _walker: Nullable<Walker>;
+
+    /**
+     * The target to be articulated by walking locomotion.
+     * When the walking locomotion feature detects walking in place, this element's
+     * X and Z coordinates will be modified to reflect locomotion. This target should
+     * be either the XR space's origin (i.e., the parent node of the WebXRCamera) or
+     * the WebXRCamera itself. Note that the WebXRCamera path will modify the position
+     * of the WebXRCamera directly and is thus discouraged.
+     */
+    public locomotionTarget: WebXRCamera | TransformNode;
+
+    /**
+     * Construct a new Walking Locomotion feature.
+     * @param sessionManager manager for the current XR session
+     * @param options creation options, prominently including the vector target for locomotion
+     */
+    public constructor(sessionManager: WebXRSessionManager, options: IWebXRWalkingLocomotionOptions) {
+        super(sessionManager);
+        this._sessionManager = sessionManager;
+        this.locomotionTarget = options.locomotionTarget;
+        if (this.locomotionTarget instanceof WebXRCamera) {
+            Logger.Warn("Using walking locomotion directly on a WebXRCamera may have unintended interactions with other XR techniques. Using an XR space parent is highly recommended");
+        }
+    }
+
+    /**
+     * Attaches the feature.
+     * Typically called automatically by the features manager.
+     * @returns true if attach succeeded, false otherwise
+     */
+    public attach(): boolean {
+        if (!super.attach()) {
+            return false;
+        }
+
+        this._walker = new Walker();
+        return true;
+    }
+
+    /**
+     * Detaches the feature.
+     * Typically called automatically by the features manager.
+     * @returns true if detach succeeded, false otherwise
+     */
+    public detach(): boolean {
+        if (!super.detach()) {
+            return false;
+        }
+
+        this._walker = null;
+        return true;
+    }
+
+    protected _onXRFrame(frame: XRFrame): void {
+        const pose = frame.getViewerPose(this._sessionManager.baseReferenceSpace);
+        if (!pose) {
+            return;
+        }
+
+        const handednessScalar = this.locomotionTarget.getScene().useRightHandedSystem ? 1 : -1;
+
+        const m = pose.transform.matrix;
+        this._up.copyFromFloats(m[4], m[5], handednessScalar * m[6]);
+        this._forward.copyFromFloats(m[8], m[9], handednessScalar * m[10]);
+        this._position.copyFromFloats(m[12], m[13], handednessScalar * m[14]);
+
+        // Compute the nape position
+        this._forward.scaleAndAddToRef(0.05, this._position);
+        this._up.scaleAndAddToRef(-0.05, this._position);
+        this._walker!.update(this._position, this._forward);
+        this._movement.copyFrom(this._walker!.movementThisFrame);
+        if (this.locomotionTarget instanceof TransformNode) {

```

Comment:
Does that future-proof against inheritance? As an idea since I'm (perhaps needlessly) worried about that, I've added a cached-off `instanceof` result so we do the slow thing only once. If you believe the string comparison thing is better, though, I'll change it to that.

Reply:
same as above, can be discussed outside of this context :-)